sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferStore.scala

```diff
@@ -304,6 +320,34 @@ abstract class RapidsBufferStore(
       }
     }
 
+    override def getDeviceMemoryBuffer: DeviceMemoryBuffer = {
+      (0 until MAX_UNSPILL_ATTEMPTS).foreach { _ =>
+        catalog.acquireBuffer(id, DEVICE) match {
+          case Some(buffer) =>
+            withResource(buffer) { buffer =>
+              return buffer.getDeviceMemoryBuffer
+            }
+          case _ =>
+            try {
+              val newBuffer = {
+                logDebug(s"Unspilling $this $id to $DEVICE")
+                RapidsBufferCatalog.getDeviceStorage.copyBuffer(
+                  this, getMemoryBuffer, Cuda.DEFAULT_STREAM)
```

Comment:
Actually this doesn't work for the GDS case as it causes an infinite regress.

Reply:
It doesn't work because this calls the wrong `getMemoryBuffer`.  This is calling `getMemoryBuffer` on the `RapidsBuffer` rather than on the buffer store method passing `this` as the argument.  In retrospect, `getMemoryBuffer` is a poor name choice on the store method because it's easy to confuse with the buffer method.  We can use the original name of `materializeMemoryBuffer` for the buffer store method so it's harder to get confused.

Also if you think it makes more sense to move it back in the RapidsBuffer that's OK with me as long as it remains a protected method.  Externally it doesn't matter much either way whether this method lives in the store or in the buffer as long as code outside of the store cannot call it.  I don't want to allow outside code to see it since it will be confusing which one they should call, and we always want the unspill behavior for external code trying to manifest a device buffer.

Reply:
Ok, added `materializeMemoryBuffer` as a protected method in `RapidsBufferBase`. Only `RapidsGdsBuffer` needs to override it.